# 4-7_Deriving new variables

## This script contains the following points

#### 0. Imports libraries and dataframes

#### 1.a. Create "price_label" column

#### 1.b. Create "busiest_day" column

#### 2. Create "busiest_days" column

#### 3. Check "busiest_days" for accuracy

#### 4. Create "most_orders" column

#### 5. Print frequency "most_orders"

#### 6. I have a pretty code here 

#### 7. Export data as pickle file


## 0. Imports libraries and dataframes

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Importing the data frames

# creates a path with the source folder
path = r'C:\Users\Marc\Mariana_CareerFoundry'

# importing the orders.csv file from prepared data
df_ords_prods = pd.read_pickle(os.path.join(path, '02_Data','2-2_Prepared', 'ords_prods_merge.pkl'))

In [3]:
# Creating a smaller subset

df = df_ords_prods[:1000000] 
df.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2539329        1    prior             1                   2   
2   2539329        1    prior             1                   2   
3   2539329        1    prior             1                   2   
4   2539329        1    prior             1                   2   
5   2398795        1    prior             2                   3   
6   2398795        1    prior             2                   3   
7   2398795        1    prior             2                   3   
8   2398795        1    prior             2                   3   
9   2398795        1    prior             2                   3   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   
5                  7                    15.0         196                  1   
6                  7                    15.0       10258                  2   
7                  7                    15.0       12427                  3   
8                  7                    15.0       13176                  4   
9                  7                    15.0       26088                  5   

   reordered  Unnamed: 0                             product_name  aisle_id  \
0          0         195                                     Soda        77   
1          0       14084  Organic Unsweetened Vanilla Almond Milk        91   
2          0       12427                      Original Beef Jerky        23   
3          0       26089               Aged White Cheddar Popcorn        23   
4          0       26406         XL Pick-A-Size Paper Towel Rolls        54   
5          1         195                                     Soda        77   
6          0       10258                               Pistachios       117   
7          1       12427                      Original Beef Jerky        23   
8          0       13176                   Bag of Organic Bananas        24   
9          1       26089               Aged White Cheddar Popcorn        23   

   department_id  prices _merge  
0              7     9.0   both  
1             16    12.5   both  
2             19     4.4   both  
3             19     4.7   both  
4             17     1.0   both  
5              7     9.0   both  
6             19     3.0   both  
7             19     4.4   both  
8              4    10.3   both  
9             19     4.7   both

## 1.a. Create "price_label" column

In [4]:
# Creates a "price_range" FUNCTION

def price_range(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'


In [5]:
# Creates a "price_label" COLUMN

df['price_label'] = df.apply(price_range, axis=1)

C:\Users\Marc\AppData\Local\Temp\ipykernel_16180\310669160.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_label'] = df.apply(price_range, axis=1)


In [6]:
# Checks outputs of "price_label" COLUMN

df['price_label'].value_counts(dropna = False)

price_label
Mid-range product    673183
Low-range product    314392
High range            12425
Name: count, dtype: int64

## 1.b. Create "busiest_day" column

In [7]:
# Creates "busiest_day" column

busiest_day = []

for value in df["orders_day_of_week"]:
  if value == 0:
    busiest_day.append("Busiest day")
  elif value == 4:
    busiest_day.append("Least busy")
  else:
    busiest_day.append("Regularly busy")
    
df['busiest_day'] = busiest_day

C:\Users\Marc\AppData\Local\Temp\ipykernel_16180\2217561161.py:13: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['busiest_day'] = busiest_day


In [8]:
# Checks the busiest day column

df['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    692455
Busiest day       189676
Least busy        117869
Name: count, dtype: int64

## 2. Create "busiest_days" column

In [9]:
# Look for the frequency distribution

df['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    189676
1    169888
6    139752
5    133956
2    130531
3    118328
4    117869
Name: count, dtype: int64

#### Busiest days ranking

1. 0 = Saturday
2. 1 = Sunday
3. 6 = Friday
4. 5 = Thursday
5. 2 = Monday
6. 3 = Tuesday
7. 4 = Wednesday

In [10]:
# Creates "busiest_days" column

busiest_days = []

for value in df["orders_day_of_week"]:
  if value == 0:
    busiest_days.append("Busiest days") 
  elif value == 1:
    busiest_days.append("Busiest days")    
  elif value == 4:
    busiest_days.append("Least busy days")
  elif value == 3:
    busiest_days.append("Least busy days")
  else:
    busiest_days.append("Normal busy")
    
df['busiest_days'] = busiest_days

C:\Users\Marc\AppData\Local\Temp\ipykernel_16180\3403341039.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['busiest_days'] = busiest_days


## 3. Check "busiest_days" for accuracy

In [11]:
# Checks the busiest day column

df['busiest_days'].value_counts(dropna = False)

busiest_days
Normal busy        404239
Busiest days       359564
Least busy days    236197
Name: count, dtype: int64

### Checking Outputs

#### Busiest day = Saturday (189676) + Sunday (169888) = 359564 - Ok!

#### Least busy days = Tuesday (118328) + Wednesday (117869) = 236197 - Ok!

## 4. Create "most_orders" column

In [12]:
# Look for the frequency distribution

df['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    86480
11    84828
13    81566
14    81389
15    81169
12    80821
9     77729
16    76743
17    65219
8     53867
18    51326
19    39413
20    29683
7     27843
21    24565
22    19270
23    11955
6      9542
0      5962
1      2934
5      2683
2      1867
4      1664
3      1482
Name: count, dtype: int64

### Analysis: very conveniently, the most 6 busy hours are sequential from 10 to 15 and the 6 least busy houras are also sequential from 0 to 5. I can write than a simpler command using intervals 

### Using loc()

In [13]:
df.loc[df['order_hour_of_day'] <= 5, 'busy_hours_loc'] = 'Fewest orders'

C:\Users\Marc\AppData\Local\Temp\ipykernel_16180\1950201652.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['order_hour_of_day'] <= 5, 'busy_hours_loc'] = 'Fewest orders'


In [14]:
df.loc[(df['order_hour_of_day'] >= 6) & (df['order_hour_of_day'] <= 9), 'busy_hours_loc'] = 'Average orders'

In [15]:
df.loc[(df['order_hour_of_day'] >= 10) & (df['order_hour_of_day'] <= 15), 'busy_hours_loc'] = 'Most orders'

In [16]:
df.loc[(df['order_hour_of_day'] >= 16) & (df['order_hour_of_day'] <= 23), 'busy_hours_loc'] = 'Average orders'

In [17]:
df['busy_hours_loc'].value_counts(dropna = False)

busy_hours_loc
Most orders       496253
Average orders    487155
Fewest orders      16592
Name: count, dtype: int64

### Using for loop

In [18]:
# Creates "busiest_days" column

busiest_hours = []

for value in df["order_hour_of_day"]:

  if value <= 5:
    busiest_hours.append("Fewest orders")
  elif value <= 15 & value > 9 :
    busiest_hours.append("Most orders") 
  else:
    busiest_hours.append("Average orders")

In [19]:
df['busiest_period_of_day'] = busiest_hours

C:\Users\Marc\AppData\Local\Temp\ipykernel_16180\2025454875.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['busiest_period_of_day'] = busiest_hours


## 5. Print frequency "most_orders"

In [20]:
# Checks the busiest period_of_day column

df['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       496253
Average orders    487155
Fewest orders      16592
Name: count, dtype: int64

## 7. Export data as pickle file

In [21]:
# Exports data to pkl

df.to_pickle(os.path.join(path, '02_Data','2-2_Prepared', 'short_analysis.pkl'))